File: ScopusCrawler/ScopusCrawler/Processing/Worker.cs
using ScopusCrawler.Models;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Threading;
using System.Xml.Linq;
using System.Collections.Concurrent;

namespace ScopusCrawler.Processing
{
    public class Worker
    {
        string ApiUrl = "http://api.elsevier.com/content/search/scopus?sort=citedby-count&httpAccept=application/xml&";
        int ResultsPerPage = 25;
        BlockingCollection<string> brokenArticles;
        BlockingCollection<string> processedArticles;

        public Worker(BlockingCollection<string> articlesProcessed, BlockingCollection<string> articlesBroken)
        {
            processedArticles = articlesProcessed;
            brokenArticles = articlesBroken;
        }

        public List<Article> Start(int year, string journalISSN)
        {
            var query = "query=ISSN(" + journalISSN + ")";
            var initRequest = ApiUrl + "&" + query + "&date=" + year;
            var initData = GetData(initRequest);
            var doc = XElement.Parse(initData);
            var totalResultsNumber = int.Parse(doc.Elements().First().Value);
            if (totalResultsNumber > 4975)
                totalResultsNumber = 4975;
            var splitsCount = totalResultsNumber / ResultsPerPage;
            if (totalResultsNumber % ResultsPerPage != 0)
            {
                splitsCount++;
            }
            var result = new List<Article>();
            for (int i = 0; i < splitsCount; i++)
            {
                var request = ApiUrl + "&" + query + "&date=" + year + "&start=" + (i * 25).ToString();
                var data = GetData(request);
                if (data != string.Empty)
                {
                    var element = XElement.Parse(data);
                    var articleElements = element.Elements(XName.Get("{http://www.w3.org/2005/Atom}entry"));
                    var articles = CollectArticles(articleElements);
                    result.AddRange(articles);
                    //foreach (var articleElement in articleElements)
                    //{
                    //    try
                    //    {
                    //        var article = new Article();
                    //        var id = articleElement.Element(XName.Get("{http://www.w3.org/2005/Atom}identifier")).Value;
                    //        article.ScopusID = id.Replace("SCOPUS_ID:", "");
                    //        GetArticleData(ref article);
                    //    }
                    //    catch (Exception ex)
                    //    {
                    //        Console.WriteLine(ex.Message);
                    //    }
                    //}
                }
            }
            return result;
        }

        List<Article> CollectArticles(IEnumerable<XElement> articleElements)
        {
            var result = new List<Article>();
            foreach (var articleElement in articleElements)
            {
                try
                {
                    var article = new Article();
                    var id = articleElement.Element(XName.Get("{http://www.w3.org/2005/Atom}identifier")).Value;
                    article.ScopusID = id.Replace("SCOPUS_ID:", "");
                    GetArticleData(ref article);
                    result.Add(article);
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex.Message);
                }
            }
            return result;
        }

        void GetArticleData(ref Article article)
        {
            //using (ScopusDbContext dbContext = new ScopusDbContext())
            {
                {
                    var articleId = article.ScopusID;
                    if (processedArticles.Contains(articleId))
                    {
                        return;
                    }
                    Console.WriteLine("Working on article " + article.ScopusID);
                    var apiUrl = "http://api.elsevier.com/content/abstract/scopus_id/{0}?httpAccept=application/xml";
                    var request = string.Format(apiUrl, article.ScopusID);
                    var data = GetData(request);
                    if (data != string.Empty)
                    {
                        var element = XElement.Parse(data);
                        var coreElement = element.Element(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}coredata"));
                        try
                        {
                            article.Title = coreElement.Element(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}title")).Value;
                            article.PublicationDate = coreElement.Element(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}coverDate")).Value;
                        }
                        catch
                        {
                            brokenArticles.Add(article.ScopusID);
                            return;
                        }
                        article.CitationsCount = int.Parse(coreElement.Element(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}citedby-count")).Value);
                        var descriptionElement = coreElement.Elements(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}description"));
                        var abstractsElement = descriptionElement.Elements(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}abstract"));
                        if (abstractsElement.Count() > 0)
                        {
                            article.Abstracts = abstractsElement.First().Value;
                        }
                        var affiliationsElement = element.Elements(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}affiliation"));
                        var affiliations = GetAffiliations(affiliationsElement);
                        var authorsElement = element.Element(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}authors"));
                        article.Authors = GetAuthors(authorsElement, affiliations);
                        var doiElement = coreElement.Elements(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}doi"));
                        if (doiElement.Count() > 0)
                        {
                            article.DOI = doiElement.First().Value;
                        }
                        var keywordsElement = element.Elements(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}authkeywords"));
                        foreach (var keyword in keywordsElement.Elements())
                        {
                            article.Keywords.Add(keyword.Value);
                        }
                        var recordsItem = element.Element(XName.Get("item"));
                        var bibrecordItem = recordsItem.Element(XName.Get("bibrecord"));
                        var tailItem = bibrecordItem.Element(XName.Get("tail"));
                        var subjectAreasElement = element.Element(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}subject-areas"));
                        article.SubjectAreas = GetSubjectAreas(subjectAreasElement);
                        var referencesItem = tailItem.Element(XName.Get("bibliography"));
                        if (referencesItem != null)
                        {
                            article.References = GetReferences(referencesItem.Elements());
                        }
                        //if (!dbContext.Articles.Any(a => a.ScopusID == articleId))
                        //    dbContext.Articles.Add(article);
                        //dbContext.SaveChanges();
                        //Console.WriteLine("Added article: " + article.Title);
                        //processedArticles.Add(article.ScopusID);
                    }
                }
            }
        }

        private List<SubjectArea> GetSubjectAreas(XElement subjectAreasElement)
        {
            var result = new List<SubjectArea>();
            //using (ScopusDbContext dbContext = new ScopusDbContext())
            {
                foreach (var element in subjectAreasElement.Elements())
                {
                    var area = new SubjectArea()
                    {
                        ScopusID = element.Attribute(XName.Get("code")).Value,
                        Abbreviation = element.Attribute(XName.Get("abbrev")).Value,
                        Name = element.Value
                    };
                    //var subjectArea = dbContext.SubjectAreas.FirstOrDefault(a => a.ScopusID == area.ScopusID);
                    //if (subjectArea == null)
                    //{
                    //    dbContext.SubjectAreas.Add(area);
                    //    dbContext.SaveChanges();
                    //    subjectArea = area;
                    //    Console.WriteLine("Added subject area: " + area.Name);
                    //}
                    //result.Add(subjectArea);
                    result.Add(area);
                }
            }
            return result;
        }

        private List<Article> GetReferences(IEnumerable<XElement> referencesElements)
        {
            var result = new List<Article>();
            //using (ScopusDbContext dbContext = new ScopusDbContext())
            {
                foreach (var element in referencesElements)
                {
                    var refInfoElement = element.Element(XName.Get("ref-info"));
                    var refIdListElement = refInfoElement.Element(XName.Get("refd-itemidlist"));
                    foreach (var refIdElement in refIdListElement.Elements())
                    {
                        var refType = refIdElement.Attribute(XName.Get("idtype")).Value;
                        if (refType == "SGR")
                        {
                            var refId = refIdElement.Value;
                            var refArticle = new Article();
                            refArticle.ScopusID = refId;
                            //if (!brokenArticles.Contains(refId) && !processedArticles.Contains(refId))
                            //{
                            //    if (!dbContext.Articles.Any(a => a.ScopusID == refId))
                            //    {
                            //        var refArticle = new Article();
                            //        refArticle.ScopusID = refId;
                            //        dbContext.Articles.Add(refArticle);
                            //        dbContext.SaveChanges();
                            //        //GetArticleData(ref refArticle);
                            //    }
                            //    try
                            //    {
                            //        result.Add(dbContext.Articles.First(a => a.ScopusID == refId));
                            //    }
                            //    catch
                            //    {
                            //        Console.WriteLine("--- Broken reference! ---");
                            //    }
                            //}
                            result.Add(refArticle);
                        }
                    }
                }
            }
            return result;
        }

        private List<Affiliation> GetAffiliations(IEnumerable<XElement> affiliationElements)
        {
            var result = new List<Affiliation>();
            //using (ScopusDbContext dbContext = new ScopusDbContext())
            {
                foreach (var element in affiliationElements)
                {
                    var scopusId = element.Attribute(XName.Get("id")).Value;
                    //var affiliation = dbContext.Affiliations.FirstOrDefault(a => a.ScopusID == scopusId);
                    //if (affiliation == null)
                    //{
                        var affiliation = new Affiliation()
                        {
                            ScopusID = scopusId,
                            Name = element.Element(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}affilname")).Value,
                            City = element.Element(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}affiliation-city")).Value,
                            Country = element.Element(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}affiliation-country")).Value
                        };
                        //dbContext.Affiliations.Add(affiliation);
                        //dbContext.SaveChanges();
                    //    Console.WriteLine("Added affiliation: " + affiliation.Name);
                    //}
                    result.Add(affiliation);
                }
            }
            return result;
        }

        private List<Author> GetAuthors(XElement authorsElement, List<Affiliation> affiliations)
        {
            var result = new List<Author>();
            //using (ScopusDbContext dbContext = new ScopusDbContext())
            {
                foreach (var element in authorsElement.Elements())
                {
                    var scopusId = element.Attribute(XName.Get("auid")).Value;
                    //var author = dbContext.Authors.FirstOrDefault(a => a.ScopusID == scopusId);
                    //if (author == null)
                    //{
                        var author = new Author();
                        author.ScopusID = scopusId;
                        var fullNameElement = element.Element(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}indexed-name"));
                        if (fullNameElement != null)
                            author.FullName = fullNameElement.Value;
                        var initialsElement = element.Element(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}initials"));
                        if (initialsElement != null)
                            author.Initials = initialsElement.Value;
                        var surnameElement = element.Element(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}surname"));
                        if (surnameElement != null)
                            author.Surname = surnameElement.Value;
                        var affiliationElement = element.Element(XName.Get("{http://www.elsevier.com/xml/svapi/abstract/dtd}affiliation"));
                        if (affiliationElement != null)
                        {
                            var affiliationId = affiliationElement.Attribute(XName.Get("id")).Value;
                            var affiliation = affiliations.FirstOrDefault(a => a.ScopusID == affiliationId);
                            if (affiliation != null)
                                author.Affiliation = affiliations.First(a => a.ScopusID == affiliationId);
                            else
                                author.AffiliationId = 240;
                        }
                        else
                        {
                            author.AffiliationId = 240;
                        }
                        //dbContext.Authors.Add(author);
                        //dbContext.SaveChanges();
                        //Console.WriteLine("Added author: " + author.FullName);
                    //}
                    result.Add(author);
                }
            }
            return result;
        }

        string GetData(string url)
        {
            try
            {
                using (ScopusDbContext dbContext = new ScopusDbContext())
                {
                    var requestDone = dbContext.Requests.FirstOrDefault(r => r.Request == url);
                    if (requestDone == null || requestDone.Response == string.Empty)
                    {
                        var key = KeysStorage.GetKey();
                        var requestUrl = url + "&apiKey=" + key;
                        var request = (HttpWebRequest)WebRequest.Create(requestUrl);
                        request.Proxy = new WebProxy("proxy.ifmo.ru", 3128);
                        var response = request.GetResponse();
                        Stream dataStream = response.GetResponseStream();
                        StreamReader reader = new StreamReader(dataStream);
                        string responseString = reader.ReadToEnd();
                        responseString = responseString.Replace("prism:", "").Replace("dc:", "").Replace("opensearch:", "").Replace("ce:", "").Replace("atom:", "");
                        dbContext.Requests.Add(new RequestDone()
                        {
                            Request = url,
                            Response = responseString
                        });
                        dbContext.ChangeTracker.DetectChanges();
                        dbContext.SaveChanges();
                        dbContext.Dispose();
                        //Thread.Sleep(1000);
                        return responseString;
                    }
                    else
                    {
                        return requestDone.Response;
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("Request failed.");
                return string.Empty;
            }
        }
    }
}

File: crawler/worker.go
package crawler

import (
	"errors"
	"strconv"
	"strings"

	uuid "github.com/satori/go.uuid"
	"github.com/visheratin/scopus-crawler/config"
	"github.com/visheratin/scopus-crawler/logger"
	"github.com/visheratin/scopus-crawler/models"
	"github.com/visheratin/scopus-crawler/query"
	"github.com/visheratin/scopus-crawler/storage"
)

type Worker struct {
	Config      config.Configuration
	Storage     storage.GenericStorage
	DataSources []DataSource
	Work        chan SearchRequest
	WorkerQueue chan chan SearchRequest
}

func (worker *Worker) Start() {
	go func() {
		for {
			worker.WorkerQueue <- worker.Work
			work := <-worker.Work
			ds := work.Source
			switch work.SourceName {
			case "search":
				data, err := query.MakeQuery(ds.Path, "", work.Fields, worker.Config.RequestTimeout, worker.Storage)
				if err != nil {
					logger.Error.Println(err)
					return
				}
				articles, err := ExtractArticles(data)
				if err != nil {
					logger.Error.Println(err)
					return
				}
				var articleDs DataSource
				for _, v := range worker.DataSources {
					if v.Name == "article" {
						articleDs = v
						break
					}
				}
				for _, article := range articles {
					err := worker.ProceedArticle(article, articleDs, 0)
					if err != nil {
						logger.Error.Println("Error on proceeding article with id=" + article.ScopusID)
						logger.Error.Println(err)
					}
				}
			}
		}
	}()
}

func ExtractArticles(rawResponse map[string]interface{}) ([]models.Article, error) {
	result := []models.Article{}
	searchContainer, searchSucceed := rawResponse["search-results"]
	if searchSucceed {
		rawEntries, ok := searchContainer.(map[string]interface{})["entry"]
		if !ok {
			return result, errors.New("error on parsing search-results element")
		}
		entries, ok := rawEntries.([]interface{})
		if !ok {
			return result, errors.New("error on parsing entries element")
		}
		for _, value := range entries {
			entry, ok := value.(map[string]interface{})
			if !ok {
				continue
			}
			id := entry["dc:identifier"].(string)
			scopusID := strings.Replace(id, "SCOPUS_ID:", "", -1)
			article := models.Article{
				ScopusID: scopusID,
			}
			title, ok := entry["dc:title"]
			if ok {
				article.Title = title.(string)
			}
			abstracts, ok := entry["dc:description"]
			if ok {
				article.Abstracts = abstracts.(string)
			}
			citations, ok := entry["citedby-count"]
			if ok {
				count, err := strconv.Atoi(citations.(string))
				if err == nil {
					article.CitationsCount = count
				}
			}
			pubDate, ok := entry["prism:coverDate"]
			if ok {
				article.PublicationDate = pubDate.(string)
			}
			pubType, ok := entry["prism:aggregationType"]
			if ok {
				article.PublicationType = pubType.(string)
			}
			pubName, ok := entry["prism:publicationName"]
			if ok {
				article.PublicationTitle = pubName.(string)
			}

			article.Authors = []models.Author{}
			authorsElem := entry["author"]
			authors := authorsElem.([]interface{})
			for _, authorVal := range authors {
				authorElem := authorVal.(map[string]interface{})
				author := models.Author{}
				id, ok := authorElem["authid"]
				if ok {
					author.ScopusID = id.(string)
				}
				name, ok := authorElem["authname"]
				if ok {
					author.IndexedName = name.(string)
				}
				surname, ok := authorElem["surname"]
				if ok {
					author.Surname = surname.(string)
				}
				firstname, ok := authorElem["given-name"]
				if ok {
					author.Name = firstname.(string)
				}
				author.Affiliation = models.Affiliation{}
				authAffElem := authorElem["afid"]
				authAff := authAffElem.([]interface{})
				if len(authAff) > 0 {
					affID, ok := authAff[0].(map[string]interface{})["$"]
					if ok {
						author.AffiliationID = affID.(string)
					}
				}
				article.Authors = append(article.Authors, author)
			}

			article.Affiliations = []models.Affiliation{}
			affElem := entry["affiliation"]
			affiliations := affElem.([]interface{})
			for _, affVal := range affiliations {
				affElem := affVal.(map[string]interface{})
				affiliation := models.Affiliation{}
				id, ok := affElem["afid"]
				if ok {
					affiliation.ScopusID = id.(string)
				}
				title, ok := affElem["affilname"]
				if ok {
					affiliation.Title = title.(string)
				}
				city, ok := affElem["affiliation-city"]
				if ok {
					affiliation.City = city.(string)
				}
				country, ok := affElem["affiliation-country"]
				if ok {
					affiliation.Country = country.(string)
				}
				article.Affiliations = append(article.Affiliations, affiliation)
			}

			result = append(result, article)
		}
		return result, nil
	}
	return []models.Article{}, errors.New("empty search response")
}

func (worker *Worker) ProceedArticle(article models.Article, articleDs DataSource, depth int) error {
	articleData, err := query.MakeQuery(articleDs.Path, article.ScopusID, map[string]string{}, worker.Config.RequestTimeout, worker.Storage)
	if err != nil {
		logger.Error.Println("Error on requesting data for id=" + article.ScopusID)
		logger.Error.Println(err)
	}
	articleContainer, articleRetrieved := articleData["abstracts-retrieval-response"]
	if articleRetrieved {
		article.Keywords = []models.Keyword{}
		keywordsContainer, ok := articleContainer.(map[string]interface{})["authkeywords"]
		if !ok {
			return errors.New("error on parsing authkeywords element")
		}
		keywordsVal, ok := keywordsContainer.(map[string]interface{})["author-keyword"]
		if !ok {
			return errors.New("error on parsing author-keyword element")
		}
		keywords, ok := keywordsVal.([]interface{})
		if !ok {
			return errors.New("error on parsing keywords element")
		}
		for _, keywordVal := range keywords {
			keyword := models.Keyword{}
			keyword.ID = uuid.NewV4().String()
			keywordElem := keywordVal.(map[string]interface{})
			value, ok := keywordElem["$"]
			if ok {
				keyword.Value = value.(string)
				article.Keywords = append(article.Keywords, keyword)
			}
		}

		article.SubjectAreas = []models.SubjectArea{}
		areaContainer, ok := articleContainer.(map[string]interface{})["subject-areas"]
		if !ok {
			return errors.New("error on parsing subject-areas element")
		}
		areasVal, ok := areaContainer.(map[string]interface{})["subject-area"]
		if !ok {
			return errors.New("error on parsing subject-area element")
		}
		areas, ok := areasVal.([]interface{})
		if !ok {
			return errors.New("error on parsing subject areas element")
		}
		for _, areaVal := range areas {
			area := models.SubjectArea{}
			areaElem := areaVal.(map[string]interface{})
			name, ok := areaElem["$"]
			if ok {
				area.Description = name.(string)
			}
			code, ok := areaElem["@code"]
			if ok {
				area.Code = code.(string)
			}
			abbrev, ok := areaElem["@abbrev"]
			if ok {
				area.Title = abbrev.(string)
			}
			article.SubjectAreas = append(article.SubjectAreas, area)
		}

		article.References = []models.Article{}
		refContainer, ok := articleContainer.(map[string]interface{})["item"]
		if !ok {
			return errors.New("error on parsing item element")
		}
		refContainer, ok = refContainer.(map[string]interface{})["bibrecord"]
		if !ok {
			return errors.New("error on parsing bibrecord element")
		}
		refContainer, ok = refContainer.(map[string]interface{})["tail"]
		if !ok {
			return errors.New("error on parsing tail element")
		}
		refContainer, ok = refContainer.(map[string]interface{})["bibliography"]
		if !ok {
			return errors.New("error on parsing bibliography element")
		}
		refContainer, ok = refContainer.(map[string]interface{})["reference"]
		if !ok {
			return errors.New("error on parsing reference element")
		}
		references, ok := refContainer.([]interface{})
		if !ok {
			return errors.New("error on parsing references element")
		}
		for _, refVal := range references {
			reference := models.Article{}
			refElem := refVal.(map[string]interface{})
			refData, ok := refElem["ref-info"]
			if !ok {
				continue
			}
			refData, ok = refData.(map[string]interface{})["refd-itemidlist"]
			if !ok {
				continue
			}
			refData, ok = refData.(map[string]interface{})["itemid"]
			if !ok {
				continue
			}
			refElem = refData.(map[string]interface{})
			idType, ok := refElem["@idtype"]
			if ok && idType == "SGR" {
				id, ok := refElem["$"]
				if ok {
					reference.ScopusID = id.(string)
					if depth < worker.Config.ReferencesDepth {
						worker.ProceedArticle(reference, articleDs, depth+1)
					}
					article.References = append(article.References, reference)
				}
			}
		}
		for _, reference := range article.References {
			err := worker.Storage.CreateArticle(reference)
			if err != nil {
				return err
			}
		}
		err := worker.Storage.CreateArticle(article)
		if err != nil {
			return err
		}
		return nil
	}
	return errors.New("empty article response")
}
